packages/node/src/utils/kyve/kyve.ts

```diff
@@ -0,0 +1,413 @@
+
+      const fromKey = parseDecimal(midBundle.from_key);
+      const toKey = parseDecimal(midBundle.to_key);
+
+      if (height >= fromKey && height <= toKey) {
+        startBundleId = mid;
+        this.currentBundleId = startBundleId;
+        return startBundleId;
+      }
+
+      if (height > toKey) {
+        low = mid + 1;
+      } else {
+        high = mid - 1;
+      }
+    }
+    throw new Error(`No suitable bundle found for height ${height}}`);
+  }
+
+  private findBlockByHeight(
+    height: number,
+    fileCacheData: UnZippedKyveBlockReponse[],
+  ): UnZippedKyveBlockReponse {
+    return fileCacheData.find(
+      (bk: UnZippedKyveBlockReponse) => bk.key === height.toString(),
+    );
+  }
+
+  async updateCurrentBundleAndDetails(
+    height: number,
+  ): Promise<UnZippedKyveBlockReponse[]> {
+    // this is on init, and then when height is greater than current cache
+    if (
+      this.currentBundleId === -1 ||
+      parseDecimal(this.cachedBundleDetails.to_key) < height
+    ) {
+      this.currentBundleId =
+        this.currentBundleId === -1
+          ? await this.getBundleId(height)
+          : this.currentBundleId + 1;
+      this.cachedBundleDetails = await this.getBundleById(this.currentBundleId);
+      try {
+        await fs.promises.access(
+          path.join(
+            this.tmpCacheDir,
+            `bundle_${parseDecimal(this.cachedBundleDetails.id) - 2}`,
+          ),
+        );
+        await this.clearFileCache();
+      } catch (e) {
+        /* empty */
+      }
+
+      return this.jsonParseWrapper(await this.getFileCacheData());
+    }
+  }
+
+  async pollUntilReadable(bundleFilePath: string): Promise<string> {
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      try {
+        return await this.readFromFile(bundleFilePath);
+      } catch (e) {
+        if (e.code === 'EACCES') {
+          await delay(POLL_TIMER);
+        } else {
+          throw e;
+        }
+      }
+    }
+  }
+
+  async downloadAndProcessBundle(bundleFilePath: string): Promise<void> {
+    const writeStream = fs.createWriteStream(bundleFilePath, {
+      flags: 'wx+',
+    });
+
+    const zippedBundleData = await this.retrieveBundleData();
+
+    const gunzip = zlib.createUnzip({
+      maxOutputLength: MAX_COMPRESSION_BYTE_SIZE /* avoid zip bombs */,
+    });
+    zippedBundleData.data
+      .pipe(gunzip)
+      .pipe(writeStream)
+      .on('error', (err) => {
+        // TODO i am unsure if this is working with async
+        if (err.code === 'EEXIST') {
+          return this.pollUntilReadable(bundleFilePath);
+        } else {
+          throw err;
+        }
+      })
+      .on('finish', () => {
+        return fs.promises.chmod(bundleFilePath, 0o444);
+      });
+  }
+
+  async getFileCacheData(): Promise<string> {
+    const bundleFilePath = path.join(
+      this.tmpCacheDir,
+      `bundle_${this.cachedBundleDetails.id}`,
```

Comment:
Create a function for building file names. There are a couple of places they are built. Having a single place to change is the way to go.

Do they need a file extension too?